### PR TITLE
1042 Cleanup memory growing unbounded in reduce manager

### LINKED ISSUE #1042: Clear reduce state holder outer state after reduction completes
**Describe the bug**
Leaving behind data structures that need to be deleted

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #1042

### CODE REVIEW

File: src/vt/collective/reduce/reduce.impl.h [hunk 1]
```diff
@@ -291,6 +291,8 @@ void Reduce::startReduce(detail::ReduceStamp id, bool use_num_contrib) {
       theMsg()->sendMsg<MsgT,ReduceManager::reduceUp<MsgT>>(parent, typed_msg);
     }
   }
+
+  state_.erase(lookup);
```
Comment (PhilMiller) [inline src/vt/collective/reduce/reduce.impl.h:295]: Should this be here, or inside the `if (ready)` block?

Comment (lifflander) [reply src/vt/collective/reduce/reduce.impl.h:295]: Oops, good catch. I got this right on release branch, but screwed it up here.

**Comment by PhilMiller** [comment]
I've been testing locally, and this seems to fix the CI failures

**Comment by PhilMiller** [comment]
Make sure to make the same fix on the stable branch PR.

**Review by PhilMiller — APPROVED** [review]

**Comment by lifflander** [comment]
There is a reentrancy bug here due to the location of the erase.

**Review by PhilMiller — APPROVED** [review]

**Comment by lifflander** [comment]
The gcc-7 workflow keeps failing on copying a file during the install process. Is this an issue with CI? Or maybe with trace enabled we are reaching the disk usage limit on the container?

**Comment by lifflander** [comment]
Also gcc-5 has hung for nearly 4 hours. Never seen that before and can't see to access the logs right now.

**Pull request merged**